File: backend/app/schemas/transaction.py
from datetime import date as date_type
from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.models.enums import TransactionSource


class TransactionItemInput(BaseModel):
    id: UUID | None = None  # existing item to update in place (preserves added-by/on); null = new
    name: str
    quantity: Decimal = Decimal(1)
    price: Decimal
    category: str | None = None


class TransactionItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    quantity: Decimal
    price: Decimal
    category: str | None
    created_by: str | None
    updated_by: str | None
    created_at: datetime
    updated_at: datetime


class TransactionCreate(BaseModel):
    account_id: UUID | None = None
    description: str
    amount: Decimal
    currency: str | None = None
    date: date_type
    category: str | None = None
    pending: bool = False


class TransactionUpdate(BaseModel):
    # Per-transaction canonical category override; null clears it (revert to the label map/auto).
    category_override: str | None = None


class TransactionOverrideUpdate(BaseModel):
    # The caller's per-user budget overrides (in `transaction_overrides`). Only provided fields change; null
    # clears that field (revert to the account default). Toggles, never touch balances.
    include_in_spending: bool | None = None
    include_in_cash_flow: bool | None = None
    # The on-device AI category refinement to mirror server-side (provenance .aiRefined). Synced through this
    # row so other devices inherit it; distinct from the explicit `category_override`.
    refined_category: str | None = None


class TransactionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    account_id: UUID | None
    plaid_transaction_id: str | None
    # On a posted row, the pending charge's plaid id it replaced (lets the app link to a since-posted pending).
    pending_transaction_id: str | None = None
    source: TransactionSource
    description: str
    amount: Decimal
    currency: str
    date: date_type
    category: str | None
    # The caller's per-user overrides (from `transaction_overrides`); the router attaches them. null = default.
    category_override: str | None = None
    include_in_spending: bool | None = None
    include_in_cash_flow: bool | None = None
    # The on-device AI refinement mirrored from `transaction_overrides` (provenance .aiRefined). null = none.
    refined_category: str | None = None
    pending: bool
    items: list[TransactionItemResponse] = []
    created_at: datetime
    updated_at: datetime

File: ios/SplitBackAPI/Sources/SplitBackAPI/Models/Transaction.swift
import Foundation
import SwiftData

/// A bank/manual transaction. Mirrors the server `transactions` table.
/// Dedupe key for Plaid-sourced rows is `plaidTransactionId`.
///
/// Note: the type name shadows SwiftUI's `Transaction`; qualify as `SwiftUI.Transaction` where the
/// animation type is needed in views.
@Model
final class Transaction {
    @Attribute(.unique) var id: UUID
    var accountId: UUID?
    var plaidTransactionId: String?
    /// On a posted row, the `plaidTransactionId` of the pending charge it replaced (Plaid's value). Lets the
    /// app point a user from a since-posted pending transaction to its posted twin. Null otherwise.
    var pendingTransactionId: String?
    var source: TransactionSource
    var details: String
    var amount: Decimal
    var currency: String
    var date: Date
    var category: String?
    /// Explicit per-transaction canonical category (manual pick or on-device AI on this one row).
    /// Backend-synced and independent of the label-wide category map — wins over it in `effectiveCategory`.
    var categoryOverride: String?
    /// The caller's per-user budget overrides (from `transaction_overrides`); nil = the default (derive from
    /// the account). Excludes this transaction from spending / cash-flow analytics.
    var includeInSpending: Bool?
    var includeInCashFlow: Bool?
    var pending: Bool
    /// On-device (Apple Intelligence) category refinement from the merchant description, for rows whose
    /// Plaid category is vague ("Other"/uncategorized). Client-only and derived — not synced, and the
    /// transaction upsert never clears it.
    var refinedCategory: String?
    /// The on-device AI's *opinion* of this transaction's category, kept even when a higher-precedence
    /// layer wins — distinct from `refinedCategory` (a fallback used in resolution). Powers the review
    /// queue's "recategorize" suggestions. Client-only, derived, never synced; survives the upsert.
    var aiSuggestedCategory: String?
    /// Line items breaking this transaction's spend across categories (receipt itemization). Empty for a
    /// flat transaction; when present, analytics attribute each item to its own category.
    @Relationship(deleteRule: .cascade, inverse: \TransactionItem.transaction)
    var items: [TransactionItem]
    var createdAt: Date
    var updatedAt: Date

    init(
        id: UUID,
        accountId: UUID? = nil,
        plaidTransactionId: String? = nil,
        pendingTransactionId: String? = nil,
        source: TransactionSource,
        details: String,
        amount: Decimal,
        currency: String,
        date: Date,
        category: String? = nil,
        categoryOverride: String? = nil,
        includeInSpending: Bool? = nil,
        includeInCashFlow: Bool? = nil,
        pending: Bool = false,
        refinedCategory: String? = nil,
        items: [TransactionItem] = [],
        createdAt: Date,
        updatedAt: Date
    ) {
        self.id = id
        self.accountId = accountId
        self.plaidTransactionId = plaidTransactionId
        self.pendingTransactionId = pendingTransactionId
        self.source = source
        self.details = details
        self.amount = amount
        self.currency = currency
        self.date = date
        self.category = category
        self.categoryOverride = categoryOverride
        self.includeInSpending = includeInSpending
        self.includeInCashFlow = includeInCashFlow
        self.pending = pending
        self.refinedCategory = refinedCategory
        self.items = items
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}
